### PR TITLE
Advancements on offline mode, and other changes

### PR DESCRIPTION
**Pull request by JCHAVEROT**
**Associated issues: #196, #210.**

**What this PR brings:** 
      - Reworked of navigation to prevent any crashes from happening, disable Home, Maps & Records while offline.
      - Added 80% test live coverage on new code and the one for last PR.
      - Added go back button on UserProfileEditScreen composable and reduced the complexity of the class.
      - Added the possibility to edit profile info while offline, saving data so to push to the DB as soon as we refresh and have internet connection.
      - Made all previous tests pass again for CI because of the changes in navigation behavior.

**What is to be done:**
      - We still don't have to possibility to download a path and store it on internal storage. With the time constraints, I couldn't find a clean way to store the images on the internal storage that uses bytes and adapt the UI for it.

**Demo:**
CleanShot 2024-05-17 at 07 07 29

### CODE REVIEW

File: app/src/androidTest/java/com/example/triptracker/screens/OfflineScreenTest.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  @Test
+  fun testOfflineScreen() {
+    composeTestRule.onNodeWithContentDescription("Airplane Mode Active").assertIsDisplayed()
+    composeTestRule
+        .onNodeWithContentDescription("Refresh")
+        .assertIsDisplayed()
+        .assertHasClickAction()
+        .performClick()
+  }
```
Comment (cleorenaud) [inline app/src/androidTest/java/com/example/triptracker/screens/OfflineScreenTest.kt:48]: Good idea to test for the offline screen !

File: app/src/androidTest/java/com/example/triptracker/userProfile/UserProfileOverviewTest.kt [hunk 2]
```diff
@@ -9,6 +9,7 @@ import androidx.compose.ui.test.onNodeWithTag
 import androidx.compose.ui.test.onNodeWithText
```
Comment (cleorenaud) [inline app/src/androidTest/java/com/example/triptracker/userProfile/UserProfileOverviewTest.kt:9]: Looks good to me!

File: app/src/androidTest/java/com/example/triptracker/userProfile/UserProfileViewModelTest.kt [hunk 3]
```diff
@@ -217,4 +221,43 @@ class UserProfileViewModelTest {
...
+  @Test
+  fun onConnectionRefreshTest() {
+    every { mockConnection.isDeviceConnectedToInternet() } returns false
+    viewModel.tryToUpdateProfile(
+        mockNav, false, {}, Uri.EMPTY, MutableUserProfile(mutableStateOf(mockUserProfiles[0])))
+    verify { mockNav.goBack() }
+
+    every { mockConnection.isDeviceConnectedToInternet() } returns true
+    assert(viewModel.onConnectionRefresh())
+  }
```
Comment (cleorenaud) [inline app/src/androidTest/java/com/example/triptracker/userProfile/UserProfileViewModelTest.kt:249]: Well done on those tests!

File: app/src/androidTest/java/com/example/triptracker/view/NavigationBarTest.kt [hunk 4]
```diff
@@ -12,8 +12,10 @@ import androidx.compose.ui.test.onNodeWithText
 import androidx.compose.ui.test.performClick
```
Comment (cleorenaud) [inline app/src/androidTest/java/com/example/triptracker/view/NavigationBarTest.kt:12]: Looks good to me !

File: app/src/main/java/com/example/triptracker/view/Navigation.kt [hunk 5]
```diff
@@ -69,11 +69,15 @@ class Navigation(val navController: NavHostController) {
...
+  private var nextId: String? = null
 
-      if (isRetry) {
-        Log.d("Navigation", "Successfully reconnected to the internet")
+  fun navigateTo(destination: TopLevelDestination, isRetry: Boolean = false) {
+    // Check if the destination is the profile screen, which is the only destination allowed in
+    // offline mode
+    val goToProfile = destination.route == Route.PROFILE
+    if (connection.isDeviceConnectedToInternet() || goToProfile) {
+      if (isRetry || goToProfile) {
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/Navigation.kt:80]: Good idea !

File: app/src/main/java/com/example/triptracker/view/Navigation.kt [hunk 6]
```diff
@@ -103,28 +107,50 @@ class Navigation(val navController: NavHostController) {
...
+        // Pop up to the start destination of the graph to
+        // avoid building up a large stack of destinations
+
+        // on the back stack as users select items
+        popUpTo(navController.graph.findStartDestination().id) { saveState = false }
+        // Avoid multiple copies of the same destination when
+        // reselecting the same item
+        //      launchSingleTop = true
+        // Restore state when reselecting a previously selected item
+        //      restoreState = true
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/Navigation.kt:134]: Should those lines be un-commented ? Or maybe later ?

File: app/src/main/java/com/example/triptracker/view/Navigation.kt [hunk 7]
```diff
@@ -103,28 +107,50 @@ class Navigation(val navController: NavHostController) {
...
+        currentDestination = destination
+        // set the id when navigating with the map
+        navController.currentBackStackEntry?.arguments?.putString("id", id)
+        // Pop up to the start destination of the graph to
+        // avoid building up a large stack of destinations
+
+        // on the back stack as users select items
+        popUpTo(navController.graph.findStartDestination().id) { saveState = false }
+        // Avoid multiple copies of the same destination when
+        // reselecting the same item
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/Navigation.kt:131]: Lots of comments, thank you !

File: app/src/main/java/com/example/triptracker/view/NavigationBar.kt [hunk 8]
```diff
@@ -8,19 +8,24 @@ import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.unit.dp
+import com.example.triptracker.model.network.Connection
 
 @Composable
 /**
  * @param navigation (Navigation): Navigation object to navigate to other screens
  * @brief NavigationBar composable that displays the bottom navigation bar once logged in
  */
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/NavigationBar.kt:17]: We could add the @param annotation for connection

File: app/src/main/java/com/example/triptracker/view/OfflineScreen.kt [hunk 9]
```diff
@@ -35,7 +36,7 @@ import com.example.triptracker.view.theme.md_theme_orange
  * @param onRetry: Callback that is called when the user clicks on the retry button.
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/OfflineScreen.kt:36]: we might add the comment for the navigation parameter

File: app/src/main/java/com/example/triptracker/view/profile/UserProfileEditScreen.kt [hunk 10]
```diff
@@ -27,6 +27,7 @@ import androidx.compose.foundation.shape.RoundedCornerShape
 import androidx.compose.foundation.verticalScroll
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/view/profile/UserProfileEditScreen.kt:27]: Looks goo to me !

File: app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt [hunk 11]
```diff
@@ -24,8 +25,15 @@ import kotlinx.coroutines.launch
 class UserProfileViewModel(
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt:25]: We could add comments for the parameter of the view model (even though they have default values)

File: app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt [hunk 12]
```diff
@@ -245,6 +253,47 @@ class UserProfileViewModel(
     updateUserProfileInDb(profile.userProfile.value)
   }
 
+  fun onConnectionRefresh(): Boolean {
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt:256]: We could add a small comment to explain the goal of this function and what it returns

File: app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt [hunk 13]
```diff
@@ -245,6 +253,47 @@ class UserProfileViewModel(
...
+    if (isDeviceConnectedToInternet) {
+      if (profileToUpdate != null && selectedPictureToUpdate != null) {
+        updateProfile(selectedPictureToUpdate!!, profileToUpdate!!)
+      }
+      return true
+    }
+    return false
+  }
+
+  fun tryToUpdateProfile(
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt:266]: Same for this function

File: app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt [hunk 14]
```diff
@@ -245,6 +253,47 @@ class UserProfileViewModel(
...
+    if (selectedPicture == null || isDeviceConnectedToInternet) {
+      updateProfile(navigation, isCreated, onLoadingChange, selectedPicture, profile)
+    } else {
+      profileToUpdate = profile
+      selectedPictureToUpdate = Uri.parse(selectedPicture.toString())
+      navigation.goBack()
+    }
+  }
+
+  private fun updateProfile(selectedPicture: Uri, profile: MutableUserProfile) {
```
Comment (cleorenaud) [inline app/src/main/java/com/example/triptracker/viewmodel/UserProfileViewModel.kt:282]: Here too, we might benefit from some comments !

**Review by cleorenaud — APPROVED** [review]
Well done for this PR! Your code is really clean and you wrote a lot of tests.
I left you some small comments in place that might benefit from more comments but overall it is really good !

File: app/src/main/java/com/example/triptracker/view/Navigation.kt (hunk 6 above)
Comment (JCHAVEROT) [reply app/src/main/java/com/example/triptracker/view/Navigation.kt:134]: I am not the one who commented them, so I left them as-is, but I believe we can remove them

**Pull request merged**